### PR TITLE
Update to config v2, use protovalidate

### PR DESCRIPTION
**Pull request by unmultimedio**
In preparation for https://github.com/bufbuild/modules/pull/604, before publishing this module to the BSR

### CODE REVIEW

File: private/bufpkg/bufstate/global_state.go [hunk 1]
```diff
@@ -67,21 +67,3 @@ func WriteGlobalState(writer io.WriteCloser, globalState *statev1alpha1.GlobalSt
 	}
 	return nil
 }
-
-// validateGlobalState checks there are no repeated modules or empty values in names or references.
-func validateGlobalState(s *statev1alpha1.GlobalState) error {
```
Comment (unmultimedio) [inline private/bufpkg/bufstate/global_state.go:72]: This validation and the one in `module_state.go` were moved to be protovalidate rules.

File: buf.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+version: v2
```
Comment (unmultimedio) [inline buf.yaml:1]: Moved both `buf.yaml` and `buf.gen.yaml` to v2.

File: proto/state/v1alpha1/state.proto [hunk 3]
```diff
@@ -16,30 +16,40 @@ syntax = "proto3";
...
 
+import "buf/validate/validate.proto";
+
 // GlobalState is a sorted array managed modules, each one with the latest
 // reference from its local array of references. This is kept updated in a
 // global state file at the root sync directory.
 message GlobalState {
+  option (buf.validate.message).cel = {
+    id: "module_state.unique_module_names",
+    expression: "size(this.modules.filter(reference, !this.modules.map(i, i.module_name).exists_one(x, x == reference.module_name))) == 0 ? '' : (this.modules.map(reference, reference.module_name).filter(module_name, !this.modules.map(reference, reference.module_name).exists_one(x, x == module_name)))[0] + ' has appeared multiple times'"
```
Comment (unmultimedio) [inline proto/state/v1alpha1/state.proto:27]: Thx to @oliversun9

**Review by oliversun9 — APPROVED** [review]

File: buf.yaml [hunk 4]
```diff
@@ -0,0 +1,6 @@
...
+modules:
+  - path: proto
+    name: buf.build/bufbuild/modules-proto
```
Comment (unmultimedio) [inline buf.yaml:4]: This is another option

```suggestion
    name: buf.build/bufbuild/managed-modules
```

File: go.mod [hunk 5]
```diff
@@ -14,14 +17,19 @@ require (
 )
 
 require (
+	github.com/antlr4-go/antlr/v4 v4.13.0 // indirect
```
Comment (saquibmian) [inline go.mod:20]: Weird that these weren't deps from locally generated code before :/

File: private/bufpkg/bufstate/state.go [hunk 6]
```diff
@@ -114,3 +116,14 @@ func AppendModuleReference(
 	}
 	return nil
 }
+
+func validate(msg proto.Message) error {
+	v, err := protovalidate.New()
```
Comment (saquibmian) [inline private/bufpkg/bufstate/state.go:121]: Is this expensive?

File: proto/state/v1alpha1/state.proto (hunk 3 above)
Comment (saquibmian) [reply proto/state/v1alpha1/state.proto:29]: Can you document what this expression does? It's difficult to parse, I _think_ it's checking that `modules.module_name` is unique? If so, you could clarify the error message: 

> module with name %s has appeared multiple times

File: proto/state/v1alpha1/state.proto [hunk 7]
```diff
@@ -16,30 +16,42 @@ syntax = "proto3";
...
+
+option go_package = "buf.build/gen/go/bufbuild/modules-proto/protocolbuffers/go/state/v1alpha1";
+
 // GlobalState is a sorted array managed modules, each one with the latest
 // reference from its local array of references. This is kept updated in a
 // global state file at the root sync directory.
 message GlobalState {
+  option (buf.validate.message).cel = {
+    id: "module_state.unique_module_names"
+    expression: "size(this.modules.filter(reference, !this.modules.map(i, i.module_name).exists_one(x, x == reference.module_name))) == 0 ? '' : (this.modules.map(reference, reference.module_name).filter(module_name, !this.modules.map(reference, reference.module_name).exists_one(x, x == module_name)))[0] + ' has appeared multiple times'"
```
Comment (pkwarren) [inline proto/state/v1alpha1/state.proto:29]:
```suggestion
    expression: "this.modules.map(i, i.module_name).unique() ? '' : (this.modules.map(reference, reference.module_name).filter(module_name, !this.modules.map(reference, reference.module_name).exists_one(x, x == module_name)))[0] + ' has appeared multiple times'"
```

Small simplification you could make here.

File: go.mod (hunk 5 above)
Comment (unmultimedio) [reply go.mod:20]: I think these are transitive from the protovalidate deps.

File: private/bufpkg/bufstate/global_state.go (hunk 1 above)
Comment (unmultimedio) [reply private/bufpkg/bufstate/global_state.go:72]: The goal is to not have the backend depend on this module, yes, that's why we embed the validation in protovalidate rules, so we can protovalidate from the backend as well.

File: buf.yaml (hunk 4 above)
Comment (saquibmian) [reply buf.yaml:4]: CI seems to have created a `modules-proto` module, we'll want to delete that.

**Review by saquibmian — COMMENTED** [review]
LGTM, I'll wait for the small changes before approving.

File: buf.yaml (hunk 4 above)
Comment (unmultimedio) [reply buf.yaml:4]: Correct, it's the one I'm using in the backend for testing in the other PR. I'll delete it when I can replace it with this one.

File: proto/state/v1alpha1/state.proto (hunk 3 above)
Comment (unmultimedio) [reply proto/state/v1alpha1/state.proto]: Improved the error message (check the tests), and documented the CEL expression.

File: buf.yaml (hunk 4 above)
Comment (unmultimedio) [reply buf.yaml]: Renamed

File: proto/state/v1alpha1/state.proto (hunk 7 above)
Comment (unmultimedio) [reply proto/state/v1alpha1/state.proto]: Done

**Review by pkwarren — APPROVED** [review]

File: private/bufpkg/bufstate/state.go (hunk 6 above)
Comment (unmultimedio) [reply private/bufpkg/bufstate/state.go]: Done

**Review by pkwarren — APPROVED** [review]

**Review by oliversun9 — APPROVED** [review]

**Pull request merged**